### PR TITLE
DOC: attempt to explain the main different APIs

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

This is an initial attempt to explain the different Matplotlib APIs.  Note the terminology here is suggested by @timhoffm "Axes", "pyplot" and I've suggested "Data" for the third pandas/xarray interface.  

  - There are quite a few places in the docs where we could remove similar half explanations and reference this page (or what it turns into).
  - The proposed terminology should be worked into the docs in quite a few places where "OO"/"pyplot" is discussed.  

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by jklymak** [comment]
ping @mwaskom for comment and suggestions as well, as you have previously commented on the need for such a page.

**Comment by mwaskom** [comment]
Did you perhaps forget to git add the new file?

**Comment by jklymak** [comment]
Ooops!

File: doc/users/explain/api_interfaces.rst [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+This can be convenient, particularly when doing interactive work or simple
+scripts - it saves at least a line of typing, and saves the user from having
+to think about figure and axes creation.  However, for more complicated scripts
+it can be at more of a disadvantage - imagine you want to loop through a number
+of variables and plot them on separate Axes, it is often more straight-forward
+to create the Axes beforehand.  It can also become ambiguous when you have
+multiple Figures or Axes which one is being acted on.  In general the
+Matplotlib project recommends using the explicit interface, at least for code
+you want to share or preserve.
```
Comment (mwaskom) [inline doc/users/explain/api_interfaces.rst:94]: I would make this recommendation in a separate section, rather than as an aside here.

**Review by mwaskom — COMMENTED** [review]
Thanks @jklymak for starting this — I think that alleviating this very common source of confusion will be incredibly helpful for matplotlib users.

I went through and made some comments, although I didn't get to the end before I needed to get to actual work. I am not a matplotlib dev and of course you don't need my approval, but as you did ask for my feedback, I would say that this could probably be made both more focused and more accessible. IMO the current version tries to cover too many concepts and isn't as clear as it could be about the key points that it wants to make.

A couple high-level structural suggestions:

- As mentioned in the line comments, I would drop the discussion of the pandas API and the "imperative / declarative" distinction from this page.
- The primary audience for this page is probably users of the implicit interface who need some education / persuading about the explicit interface. So it might make sense to start with the implicit interface, explain what's happening "behind the scenes" in an example plot, and then show how the Axes interface makes that behind-the-scenes work explicit.
- Users would probably benefit from a structured "pros and cons" table

**Comment by jklymak** [comment]
Thanks @mwaskom very helpful comments.  I agree I wasn't sure which should go first: pyplot or "Axes".  

I would still argue that the third "data" style is _also_ confusing to folks and we end up having to deal with that confusion often enough that it is worth explaining here.  And more to the point, give folks the map back to basic Matplotlib via "ax" so they can easily see how to customize as per all the other snippets.  

If we decide to keep the third section that somewhat skews the discussion to be Axes-first though it is probably possible to do both ways.

File: doc/users/explain/api_interfaces.rst [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+  create other Artists, and build a visualization step by step in an
+  `imperative <https://en.wikipedia.org/wiki/Imperative_programming>`_
+  manner.
+- an implicit interface that keeps track of the last Figure and Axes
+  created, and adds Artists to the object it thinks the user wants. This
+  interface is also imperative.
+- a number of downstream libraries offer a
+  `declarative <https://en.wikipedia.org/wiki/Declarative_programming>`_
+  interface, usually where a data object has a ``plot`` method implemented
+  that will plot the data.
```
[the remark this answers is not in the document]
Comment (timhoffm) [reply doc/users/explain/api_interfaces.rst:25]: I think it's worthwhile to discuss this category

- This should be clearly separated from native Matplotlib think
  -  Matplotlib
      - Axes
      - pyplot
  - Third party
- I agree that we should not call it declarative. If we don't find a better name the section could be titled "How third party libraries provide plotting functions".
- Not sure if the synthetic example is easy to transfer. I'd go with the probably most used third-party lib and make a pandas example.

File: doc/users/explain/api_interfaces.rst [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+on the Axes (``plot`` in this example):
+
+.. plot::
+   :include-source:
+   :align: center
+
+    import matplotlib.pyplot as plt
+
+    fig = plt.figure()
+    ax = fig.add_axes([0.1, 0.1, 0.8, 0.8])
```
[the remark this answers is not in the document]
Comment (timhoffm) [reply doc/users/explain/api_interfaces.rst:48]: Argeed that `suplots()` is the canonical pattern and should be used here. `add_axes` is for special purposes.

**Comment by timhoffm** [comment]
@jklymak thanks for starting this. This is a long-time item on my todo list. I've collected some ideas and opinions how to present this. Still figuring out how to best contibute them here. 😄

**Comment by mwaskom** [comment]
> I would still argue that the third "data" style is also confusing to folks and we end up having to deal with that confusion often enough that it is worth explaining here. 

Yeah, I could see it being in scope to answer the question "you encourage the interface that calls methods on matplotlib objects, but how am I supposed to do that if using a third party tool?" But I wouldn't get into the (fuzzy) distinction between "imperative" / "declarative" styles or even really express it in terms of a "data interface that can plot itself." (After all, everything you would say about how pandas intersects with with the explicit / implicit interface is also true of seaborn, but that uses functions that consume a dataframe, rather than adding methods to a data container.) The important points are that there is a general pattern in multiple third-partly libraries where plotting functions/methods

- accept an `ax` kwarg, acting like the explicit interface
- otherwise grab the "current axes" and plot on it, acting like the the implicit interface
- return the `ax` object in either case, so you can further customize the plot the through the explicit interface methods

**Comment by jklymak** [comment]
Thanks @mwaskom and @timhoffm I think all your suggestions are easily actionable.

@timhoffm regarding keeping this as-is, it seems the only "permanent" decision here is whether it is in the right place in the doc hierarchy, and if the scope is approximately correct.  I'd argue we should have something that is adequate for now, and then can continue to work on it afterward.  

Wasn't there a way for us to have two code and figure blocks side-by-side at some point?   Having the "Axes" and "pyplot" examples next to each other would be cool.

File: doc/users/explain/api_interfaces.rst [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    l = ax.plot([1, 2, 3, 4], [0, 0.5, 1, 0.2])
+
+Note that ``plot`` itself also returns an object that can be subsequently
+manipulated.
+
+We call this an "explicit" interface because each object is explicitly
+referenced, and used to make the next object.  Keeping handles to the
+objects is very flexible, and allows us to customize the objects after
+they are created, but before they are displayed.
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/users/explain/api_interfaces.rst:57]: 20+ years of Matlab terminology is hard to shake...  "references" is more pythonic?

File: doc/users/explain/api_interfaces.rst [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+This is also an example of an "imperative" interface because the user tells
+Matplotlib what to do each step of the way.  Of course Matplotlib has defaults
+for many things (line widths, fonts, colors, etc.), but this interface tells
+Matplotlib how to compose the visualization.  Of course this paradigm is not
+pure, in that ``fig.add_axes()`` encapsulates creating many of the objects on
+an Axes (spines, labels, ticks) and ``ax.plot()`` encapsulates creating a
+line plot.  But it is much more "imperative" than something like
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/users/explain/api_interfaces.rst:65]: Sure, decided to drop this paragraph as not particularly useful...

**Comment by jklymak** [comment]
I can't inter sphinx xarray for some reason, but
https://65531-1385122-gh.circle-artifacts.com/0/doc/build/html/users/explain/api_interfaces.html#the-implicit-pyplot-interface

**Comment by jklymak** [comment]
Bypassed intersphinx for xarray

I'll push for this to be merged without too much more copyediting.  Our standard for doc PRs is not "is it perfect", but rather "is it an improvement".   Unless this is wrong or confusing, I think we should add it, and then others can edit to make clearer or more concise.

File: doc/users/explain/api_interfaces.rst [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+and then implements a ``plot`` method:
+
+.. plot::
+    :include-source:
+    :align: center
+
+    import matplotlib.pyplot as plt
+
+    # supplied by downstream library:
+    class DataContainer():
```
Comment (QuLogic) [inline doc/users/explain/api_interfaces.rst:206]:
```suggestion
    class DataContainer:
```

File: doc/users/explain/api_interfaces.rst [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    interface, but it also provides top-level Figure and Axes creation
+    methods, and ultimately spins up the graphical user interface, if one
+    is being used.  So ``pyplot`` is still needed regardless of the
+    interface chosen.
+
+Similarly, the declarative interfaces provided by partner libraries use the
+objects accessible by the "Axes" interface, and often accept these as arguments
+or pass them back from methods.  It is usually essential to use the explicit
+"Axes" interface to perform any customization of the default visualization, or
+to unpack the data into numpy arrays and pass directly to Matplotlib.
```
Comment (QuLogic) [inline doc/users/explain/api_interfaces.rst:263]:
```suggestion
to unpack the data into NumPy arrays and pass directly to Matplotlib.
```

**Review by QuLogic — APPROVED** [review]

**Comment by jklymak** [comment]
I'll self-merge based on the one review, but if anyone feels this does more harm than good, feel free to revert and we can re-discuss....

**Comment by QuLogic** [comment]
Oh sorry, I meant to after you had fixed the review comments.

**Comment by jklymak** [comment]
Did I miss something?

**Pull request merged**